src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionSelector.cs

```diff
@@ -234,5 +283,177 @@ protected virtual IReadOnlyList<ActionDescriptor> SelectBestActions(IReadOnlyLis
+                    {
+                        // This is a conventionally routed action - so make sure we include its 'keys' in the set of
+                        // known route value keys.
+                        conventionallyRoutedActions++;
+
+                        foreach (var kvp in action.RouteValues)
+                        {
+                            routeKeys.Add(kvp.Key);
+                        }
+                    }
+                }
+
+                // We need to hold on to an ordered set of 'keys' for the route values. We'll use these later to
+                // extract the set of route values from an incoming request to compare against our maps of known
+                // route values.
+                RouteKeys = routeKeys.ToArray();
+
+                var actionsAndKeys = new (ActionDescriptor action, string[] keys)[conventionallyRoutedActions];
+
+                var count = 0;
+                for (var i = 0; i < actions.Items.Count; i++)
+                {
+                    var action = actions.Items[i];
+                    if (action.AttributeRouteInfo == null)
+                    {
+                        // This is a conventionally routed action - so we need to extract the route values associated
+                        // with this action (in order) so we can store them in our maps.
+                        var values = new string[RouteKeys.Length];
+                        for (var j = 0; j < RouteKeys.Length; j++)
+                        {
+                            action.RouteValues.TryGetValue(RouteKeys[j], out values[j]);
+                        }
+
+                        actionsAndKeys[count++] = (action, values);
+                        
+                        if (!OrdinalIgnoreCaseEntries.TryGetValue(values, out var entries))
+                        {
+                            entries = new List<ActionDescriptor>();
+                            OrdinalIgnoreCaseEntries.Add(values, entries);
+                        }
+
+                        entries.Add(action);
+                    }
+                }
+
+                // At this point `OrdinalIgnoreCaseEntries[key].Value` contains all of the actions in the same
+                // equivalence class - meaning, all conventionally routed actions for which the route values are equal
+                // ignoring case.
+                //
+                // Now we want to iterate the actions **again** and for each unique casing, map this back to the full
+                // set for that equivalence class.
+                for (var i = 0; i < actionsAndKeys.Length; i++)
+                {
+                    var actionAndKeys = actionsAndKeys[i];
+                    if (!OrdinalEntries.ContainsKey(actionAndKeys.keys))
+                    {
+                        OrdinalEntries[actionAndKeys.keys] = OrdinalIgnoreCaseEntries[actionAndKeys.keys];
+                    }
+                }
```

Comment:
Yeah I think you're right. 👍 

This relies on the fact that the the list will be shared between the two collections. I like it.